src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs

```diff
@@ -169,173 +154,57 @@ private unsafe static OpenedReader TryOpenReaderForInMemoryPdb(IntPtr inMemoryPd
                 return null;
             }
 
-            OpenedReader result = null;
-            MetadataReaderProvider provider = null;
+            var provider = MetadataReaderProvider.FromMetadataImage((byte*)inMemoryPdbAddress, inMemoryPdbSize);
             try
             {
-                provider = MetadataReaderProvider.FromMetadataImage((byte*)inMemoryPdbAddress, inMemoryPdbSize);
-                result = new OpenedReader(provider, provider.GetMetadataReader());
+                // may throw if the metadata is invalid
+                provider.GetMetadataReader();
+                return provider;
             }
             catch (BadImageFormatException)
             {
+                provider.Dispose();
                 return null;
             }
-            finally
-            {
-                if (result == null)
-                {
-                    provider?.Dispose();
-                }
-            }
-
-            return result;
         }
 
-        private static OpenedReader TryOpenReaderFromAssemblyFile(string assemblyPath, IntPtr loadedPeAddress, int loadedPeSize)
+        private unsafe static PEReader TryGetPEReader(string assemblyPath, IntPtr loadedPeAddress, int loadedPeSize)
         {
-            Debug.Assert(assemblyPath != null);
-            Debug.Assert(loadedPeAddress != IntPtr.Zero);
-
-            // TODO: When/if PEReader supports reading loaded PE files use loadedPeAddress and loadedPeSize instead of opening the file again.
-            Stream peStream = TryOpenFile(assemblyPath);
-            if (peStream == null)
+            if (loadedPeAddress != IntPtr.Zero && loadedPeSize > 0)
             {
-                return null;
+                return new PEReader((byte*)loadedPeAddress, loadedPeSize, isLoadedImage: true);
```

Comment:
I just want to make sure this isn't going to throw an exception if the PE is invalid or something. 

During my recent SOS portable PDB work, I found that there are times when the "loadedPeAddress" actually points to a file based layout. This is when the PE is in-memory loaded from a stream (or byte[] but that isn't exposed in .NET Core anymore). Normally there is a in-memory PDB too, but there doesn't have to be. The loader is kind of a black box and the layout can depend on the platform (Windows may be different that Linux) and how the PE is loaded.  So for the SOS PDB support I had to plumb through a isFileLayout flag with the loadedPeAddress (which is the opposite sense of isLoadedImage). 

There is some work that still needs to be done on the coreclr and mscorlib StackTraceHelper side to plumb this same flag though to this code.  I just haven't had a chance to do this yet.